TimeSeriesDataTable class and DataTable class [Review 1]

klshrinidhi opened the pull request:
Files to review:
* Common/DataTable.h
* Examples/DataTable/example.cpp

TimeSeriesDataTable is ready for review. While I was working on it, I made changes to the DataTable class as well. Major ones are:
* Added iterator and proxy container support for iterating over rows and columns.
* Added more overloads to add rows/columns using containers.
* More error checking.

I have added the class TimeSeriesDataTable to the file DataTable.h itself as I think of these two classes as one.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/opensim-org/opensim-core/530)
<!-- Reviewable:end -->

chrisdembia commented:
@klshrinidhi do you have suggestions for where we should focus our review, or any points for discussion?

chrisdembia commented:
1. I still find the example file to be very intimidating. I personally would be scared from using the data table. Perhaps it can start with a really minimal "tutorial" section (with a single template type, read, write), then the more comprehensive examples that you currently have.

2. I think the name TimeSeriesDataTable is very long. I think the abbreviation `Ts` is confusing. Two ideas are `TDataTable` and `TimeSeriesTable`. My preference is `TimeSeriesTable`

chrisdembia commented on OpenSim/Examples/DataTable/example.cpp:
```diff
@@ -329,6 +369,145 @@ int main() {
+    data_real.assign(3, 20);
+    data_vec3.assign(3, SimTK::Vec3{40, 50, 60});
+    data_vec6.assign(3, SimTK::Vec6{70, 80, 90, 100, 110, 120});
+    tsdt_real.addTimestampAndRow(1.25, data_real);
```
I think this command will be used more than anything else, so it'd be nice if the name were shorter. Could this just be an overload of `addRow()`?

sherm1 replied:
> Could this just be an overload of addRow()?

Or `addRowAt(t, row)`?

chrisdembia commented on OpenSim/Examples/DataTable/example.cpp:
```diff
@@ -329,6 +369,145 @@ int main() {
+        // Get a row whose timestamp is closest in either direction to the
+        // given timestamp. Row 3 (4th row) is returned for below call.
+        auto dtrow_vec6 = tsdt_real.getRowOfTimestamp(0.5,
+                                                      LessOrGreaterThanEqual);
```
I don't know how amenable these enums are to python scripting. Even if it worked it'd feel unnatural. It seems simpler to have methods like this

```
getRowOfTimestamp
getRowBelowTimestamp
...
```

The enum could be used internally if the implementation of all those methods is very similar.

sherm1 replied:
Are there just three meaningful cases: "nearest" "at or below" and "at or above"? If so those could be good routine names.

chrisdembia commented:
Is there a convincing reason to use `timestamp` instead of `time` (e.g., `getRowAtTime()`)? Timestamp makes me think of something like `30 May 1973 18:03:05 pm UTC`, instead of a scientific notion of time. At least I think this is what timestamps are in python pandas.

sherm1 commented:
> Chris said: I think the name TimeSeriesDataTable is very long

I like that name because it is so clear, and I don't think it is too long. Since it derives from DataTable it is nice to see "DataTable" spelled out in the name. If necessary, auto or aliasing can always be used to shorten the name internally.

sherm1 commented:
> Is there a convincing reason to use timestamp instead of time

I agree that "time" would be better than "timestamp".

chrisdembia commented:
> If necessary, auto or aliasing can always be used to shorten the name internally.

Okay; maybe we can have a convention for the aliasing, like `TSDT` or `TSDataTable`, or `TsDataTable` (@klshrinidhi's choice).

I don't think a name like `TimeSeriesDataTable` is generally too long, but I do think it is too long for a class that is one of the most-used. I am fairly confident that `TimeSeriesDataTable` will be our most-used or second-most used class (perhaps second to Model).

sherm1 commented:
> Okay; maybe we can have a convention for the aliasing,

My inclination would be to wait to see whether this is a problem in real code.

sherm1 commented on OpenSim/Common/Test/DataTable.cpp:
```diff
@@ -2567,5 +2567,25 @@ int main() {
+                                         "col-four"};
+    tsdt.setColumnLabels(tsdt_labels);
+
+    tsdt.addTimestamps(std::vector<double>{1, 2, 3});
```
Will this work with just the initializer list, since that is also a container? Using "time" rather than "timestamp" it would then be
```
tsdt.addTimes({1,2,3});
```
which is very compact.

klshrinidhi replied:
There is such a method and I think I have shown it in the example code:
```
tsdt_real.addTimestamps({......});
```

sherm1 commented:
> I have added the class TimeSeriesDataTable to the file DataTable.h itself as I think of these two classes as one.

I think you should split this into DataTable.h and TimeSeriesDataTable.h. The file is 4000 lines long now and GitHub won't show me the diffs for DataTable.h so I can't do a line review.

Another thing to consider would be to separate the method declarations from their definitions, at least for the larger methods. You could move them to the end of the same header, or even into a separate header like `DataTable_Defs.h`. Either way it is easier to look at the API separately from the implementation, and you get 4 spaces of indent back in the code.

klshrinidhi commented:
* I have put the TimeSeriesDataTable class in another file.
* I will write a simpler tutorial in another PR.
* I have changed "timestamp" to "time". This took a while to do because it is not a simple search-replace.

klshrinidhi commented:
Files to review:
* Common/TimeSeriesDataTable.h
* Common/DataTable.h
* Examples/DataTable/example.cpp

sherm1 commented:
<sup>**[OpenSim/Common/DataTable.h, line 391 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHkiGzKoR9Ixlj4pcC)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/DataTable.h#L391)):</sup>
This is unnecessary if you have already determined that the iterator is eq comparable. You could just use (a==b) and !(a==b) and avoid the extra restriction. Only EqualityComparable and LessThanComparable are ever strictly required since the rest can be derived. `std::rel_ops` namespace contains generic templatized implementations of all other relational operators in terms of `==` and `<`. See discussion [here](http://www.cplusplus.com/reference/utility/rel_ops/). We should probably pull that in to each of our namespaces.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

sherm1 commented:
<sup>**[OpenSim/Common/DataTable.h, line 381 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHnRDutoDu6CPqYHBq)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/DataTable.h#L381)):</sup>
Is there a use case for `startAtColumnIndex`? It makes the method more complicated and is sometimes ignored even if supplied so I think it would be a good candidate to remove.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

sherm1 commented:
<sup>**[OpenSim/Common/DataTable.h, line 451 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHo8yxW7DmcKHEDwjm)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/DataTable.h#L451)):</sup>
Does this mean that `intializer_list` doesn't satisfy the `Container` signature above?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

sherm1 commented:
<sup>**[OpenSim/Common/DataTable.h, line 546 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHo_PEf5hqxqUWBvIp)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/DataTable.h#L546)):</sup>
There is a lot of duplication between `setColumnLabels()` and `changeColumnLabels()`. What about just redefining `setColumnLabels()` to replace existing labels if any (like any other assignment does)? Then you could remove all the `change` stuff from physical and mental space.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

sherm1 commented:
<sup>**[OpenSim/Common/DataTable.h, line 938 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHpR-BtpZDT1JyANxj)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/DataTable.h#L938)):</sup>
Trailing underscore to indicate member variables does not match the OpenSim coding guidelines, which specify leading underscore. Personally I hate leading underscore since it is borderline non-compliant with the C++ standard. Much more common (and conformant) is the `m_` convention which I have been moving to in Simbody. We should discuss to see whether the OpenSim team would be amenable to using `m_`, otherwise I think you have to stick with leading underscore. I don't think we should introduce a third convention in any case.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

sherm1 commented:
> I have put the TimeSeriesDataTable class in another file

Thanks! I still couldn't review it on GitHub for some reason but I was able to see it in Reviewable.

chrisdembia commented:
> We should discuss to see whether the OpenSim team would be amenable to using m_,

I opened a PR to discuss this.

sherm1 commented:
Review status: 0 of 4 files reviewed at latest revision, 10 unresolved discussions, some commit checks broke.

---

<sup>**[OpenSim/Common/DataTable.h, line 982 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHuneRvjvK8hpFB40N)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/DataTable.h#L982)):</sup>
Cool -- this seems really useful.

---

<sup>**[OpenSim/Common/DataTable.h, line 1024 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHtB8jKrqgflriU94F)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/DataTable.h#L1024)):</sup>
This isn't right, I think you meant !(!=). But that's unconventional anyway: you should implement `operator==`and `operator<` (if needed) and then let `std::rel_ops` do the rest.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions, some commit checks broke.

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 2 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHwT3fhqR2rNL4y5bv)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L2)):</sup>
TimeSeriesDataTable.h

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 78 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHxSHfJhfbx6bsUG9b)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L78)):</sup>
I think it would be better to restrict time to just `double`; don't make it a template argument. I think that would be sufficient for OpenSim users and less confusing for them. A more general solution could use std chrono classes providing time units, but I think that would be overkill. Other opinions?

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 88 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHypZRgsCoklsDAX3A)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L88)):</sup>
In practice, time requires at least a `double` for adequate precision, even when other parts of a simulation are done in `float`. That's because time differences are very important, and float has insufficient range for realistic delta-t sizes we encounter. This is a real problem once time gets away from 1: 100.00001 and 100.00002 shouldn't be the same value, because we often have step sizes below 1e-5.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

chrisdembia commented:
<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 78 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHxSHfJhfbx6bsUG9b)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L78)):</sup>
If we wanted a table like "IndependentVariableTable", the multiple arithmetic types definitely makes sense. But since we are restricting to time as the independent variable, I think we get to add constraints like this.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 0 of 4 files reviewed at latest revision, 18 unresolved discussions, some commit checks broke.

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 60 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuI-Lbw1Zl9stH8cgCe)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L60)):</sup>
Better to call this something like "TimesOutOfOrder".

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 156 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHzxfvuNi2IQfcM-p2)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L156)):</sup>
Should mention time complexity since users might not realize it isn't O(1).

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 179 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuI-ZtPf9MsA8ARmzq4)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L179)):</sup>
This seems like it leaves the data table in an odd state in which the number of times doesn't match the number of rows. It would be nicer if we could count on a TimeSeriesDataTable always to have times, I would think. Otherwise given one of these tables we will always have to first check if it has times before we can use it.

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 280 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuI0NVG9wKxmm-q9bQM)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L280)):</sup>
This seems good because it maintains an invariant that a TimeSeriesDataTable consists of (time,row) pairs only.

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 320 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuI0h_oQode_KDYO6JW)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L320)):</sup>
I like the `addTimesAndRows` methods, but not the separate `addTimes` stuff. Can we do without those and be able to promise that a TimeSeriesDataTable always has timed rows in it?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 0 of 4 files reviewed at latest revision, 20 unresolved discussions, some commit checks broke.

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 438 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuI1mR12ePLk6fBKceh)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L438)):</sup>
Are there use cases for these changeTime() methods? It seems like a lot of code if this isn't an important function. Maybe just have a single method that allows changing the time for a particular row? Also, one can already set all the times by creating a new TSDataTable from the underlying DataTable and a new set of times.

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 557 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuI2iz62h-CLWlIMW7Q)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L557)):</sup>
This requires an exact match (bit-identical) for the time stamp. That is kind of extreme (though probably useful on occasion). I think that requirement should be very visible, like with "ExactMatch" in the name, or using an additional NearestDir enum like "MustMatchExactly".

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

sherm1 commented:
I think there are way too many methods in the TimeSeriesDataTable class that have to do with accessing rows and elements by time. They violate the "minimal sufficiency" API design principle by a lot. And even so, they do not include what would probably be the most common access method which would be to return an interpolated row value for any time. Are we excluding interpolation from TimeSeriesDataTable intentionally?

I think we need an additional abstraction by which times are mapped to rows. Then once the desired row is known, users can use any of the existing methods for accessing the row or a single element from that row.

sherm1 commented:
I'm done with this reviewing pass. Main comments: 
- contains some great stuff, but I think there is too much of it, including a lot of duplication that could be handled instead with a "minimal sufficient" approach. 
- I would like to be able to count on a TimeSeriesDataTable invariant that its rows always have times.
- I'd like to know whether the class is supposed to deal with interpolation. Fine with me if not, but then it would be good to understand how that is supposed to work.

klshrinidhi commented:
Review status: 0 of 4 files reviewed at latest revision, 20 unresolved discussions, some commit checks broke.

---

<sup>**[OpenSim/Common/DataTable.h, line 391 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHkiGzKoR9Ixlj4pcC)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/DataTable.h#L391)):</sup>
The requirement for InputIt is that it should be at-least an InputIterator, which supports both equality and in-equality. This error checking will be replaced with 'concepts' once C++ standard comes out for it.  So I want it to ensure both operations.

---

<sup>**[OpenSim/Common/DataTable.h, line 1024 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHtB8jKrqgflriU94F)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/DataTable.h#L1024)):</sup>
Thanks. That was a bug. Fixed now.

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 78 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHxSHfJhfbx6bsUG9b)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L78)):</sup>
I talked to Ajay just now. He thinks there may be (few) cases where the timestamps will be just integers. Because there will be a type alias:
```
using TimeSeries = TimeSeriesDataTable_<SimTK::Real, SimTK::Real>
```
users will not have to worry about using incorrect types for time column in general.

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 88 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHypZRgsCoklsDAX3A)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L88)):</sup>
I thought I had it as `double`. Will change it.

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 156 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHzxfvuNi2IQfcM-p2)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L156)):</sup>
I agree. I have added the info.

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 179 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuI-ZtPf9MsA8ARmzq4)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L179)):</sup>
I have good amount of error checking in place. I have set up exceptions to be thrown when user accesses an invalid table. The time column length has to match the number of rows before user can access the table.
I talked to Ajay about it and he also thinks it is good to allow users to add rows and timestamps separately if they want to but always make sure they cannot access an invalid table.

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 438 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuI1mR12ePLk6fBKceh)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L438)):</sup>
All the overloads actually delegate the work to one function so there is not much code here.
I added these overloads for users to change the timestamps in-place (without any memory allocation) providing efficiency. Creating another DataTable will be far more expensive if they just want to change few timestamps in the time column.

If it is not going to be used much, I can mark them as advanced in doxygen so they appear at the end.

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 557 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuI2iz62h-CLWlIMW7Q)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L557)):</sup>
I think we discussed this previously in review of DataTable class. Because this is a storage class with no computation underneath, I don't check for approximate equality. I will add a comment in the documentation to make this explicit.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

sherm1 commented:
<sup>**[OpenSim/Common/DataTable.h, line 391 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHkiGzKoR9Ixlj4pcC)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/DataTable.h#L391)):</sup>
I think this came up in the last review too. From [cppreference](http://en.cppreference.com/w/cpp/concept/InputIterator):
```
The type It satisfies InputIterator if
    The type It satisfies Iterator
    The type It satisfies EqualityComparable
``` 
Concept `EqualityComparable` requires only `operator==`. But ... I think the expression `(it != it)` is supposed to work, but it doesn't have to be supported directly by the iterator class since it can be supplied locally with `std::rel_ops`. So I still think your test is too strict. 

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

sherm1 commented:
<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 78 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHxSHfJhfbx6bsUG9b)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L78)):</sup>
I think having templatized time is way overkill and we should discuss it more. It is especially awkward once we start interpolating. I think we need to revisit this in the context of how interpolating and filtering is supposed to be done on data tables.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

sherm1 commented:
<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 557 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuI2iz62h-CLWlIMW7Q)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L557)):</sup>
This is also related to interpolation. You already have methods for doing approximate matches: nearest, not greater than, not less than. So while the class itself is just for storage, interpolation is a computation. I guess this class doesn't directly support interpolation but it would be very helpful (for us reviewers) to understand how interpolation is supposed to work with it since that is such an obvious use case.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

klshrinidhi commented:
Review status: 0 of 4 files reviewed at latest revision, 20 unresolved discussions, some commit checks broke.

---

<sup>**[OpenSim/Common/DataTable.h, line 391 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHkiGzKoR9Ixlj4pcC)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/DataTable.h#L391)):</sup>
InputIterator is least strict iterator among all iterators. I will be replacing the compile time checks with the concept InputIterator at later point.

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 78 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHxSHfJhfbx6bsUG9b)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L78)):</sup>
Talking to Ajay, the idea is to have interpolation outside this class as it involves computation and this class is meant to be a storage class.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

klshrinidhi commented:
Review status: 0 of 4 files reviewed at latest revision, 20 unresolved discussions, some commit checks broke.

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 557 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuI2iz62h-CLWlIMW7Q)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L557)):</sup>
The approximate match methods just retrieve an existing row from the table. There is only comparison here. I am not sure if it can be called computation.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

sherm1 commented:
<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 179 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuI-ZtPf9MsA8ARmzq4)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L179)):</sup>
I would like to discuss this more. I don't like the idea that we can't say "a TimeSeriesDataTable is a DataTable where each row has an associated time". I guess this could be viewed as a kind of extended construction where the table is not valid until construction is complete. But that means *every* method of the TSDataTable should be checking whether the table has been completely constructed. That would be unnecessary if the class maintained a suitable invariant.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

sherm1 commented:
<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 557 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuI2iz62h-CLWlIMW7Q)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L557)):</sup>
>The approximate match methods just retrieve an existing row from the table. There is only comparison here. I am not sure if it can be called computation.

Yes, that makes sense if the class is not expected to interpolate at all. But I think its main use will be from interpolating contexts so we need to see what that will look like in order to evaluate the API. Have you already worked that out?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

klshrinidhi commented:
Review status: 0 of 4 files reviewed at latest revision, 20 unresolved discussions, some commit checks broke.

---

<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 179 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuI-ZtPf9MsA8ARmzq4)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L179)):</sup>
Discussion on this is probably a good idea.

I think (which is why I have it this way) we can say that "TimeSeriesDataTable is a DataTable with a time column". When the table is accessed, it is guaranteed to have a timestamp for each row. The table does not allow access when it is invalid. Because it allows adding rows and timestamps separately, the table is invalid (and not accessible) until the time column length matches the number of rows. 

I think the decision here is between two alternatives:
* Allow adding rows and timestamps separately. Throw when user accesses an invalid table with appropriate message. This is the current implementation.
* Do not allow the above. Only have methods that add both a row and timestamp at the same time.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

klshrinidhi commented:
Review status: 0 of 4 files reviewed at latest revision, 20 unresolved discussions, some commit checks broke.

---

<sup>**[OpenSim/Examples/DataTable/example.cpp, line 412 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHek8pcK1qBrHIyfcv-r1-412)** ([raw file](https://github.com/opensim-org/opensim-core/blob/3126fb3ab70f77adaf33052d253a67cb9b85dce2/OpenSim/Examples/DataTable/example.cpp#L412)):</sup>
Will do.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

chrisdembia commented:
<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 78 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHxSHfJhfbx6bsUG9b)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L78)):</sup>
(btw I like the TimeSeries alias a lot)

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

sherm1 commented:
<sup>**[OpenSim/Common/TimeSeriesDataTable.h, line 78 \[r2\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/530#-JuHxSHfJhfbx6bsUG9b)** ([raw file](https://github.com/opensim-org/opensim-core/blob/35b043b3898cd3a9cef03ba5ca5118f04ae4aece/OpenSim/Common/TimeSeriesDataTable.h#L78)):</sup>
> there may be (few) cases where the timestamps will be just integers.

This still does not require templatizing time! A double is a perfectly good way for storing integer times also, and that would avoid any later weirdness with interpolation since even if the times are 1,2,3 it still makes sense to ask for time 2.5.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/opensim-org/opensim-core/530)
<!-- Sent from Reviewable.io -->

chrisdembia commented:
I have an important point to make. I agree that all this templated stuff is useful but I think we should be very wary. Using templated classes makes doxygen really hard to understand; maybe not so much for us but **definitely** for our users. Look at Sets. It's really hard to get documentation for set classes because they are templates. Also consider Simbody's `Vec3`. The only documentation for it is a typedef in doxygen. If the class that is going to get heavy usage is simply a typedef, then doxygen will only show it as a typedef and users will not know to go to `TSDT<X,T>` for the actual doxygen documentation.

I wish doxygen could automatically fill out documentation for a template instance, but it doesn't seem to do so.

sherm1 commented:
Good point. I think the important thing is to consider the user while designing the class and creating its documentation. A class that can't be documented won't be of much use to our API users (could still be useful internally). It is important to sketch out some doxygen for the class, generate the documentation, and take a look at it from a user's perspective before it is too late to change anything.

chrisdembia commented:
We might get better doxygen output if we do something like:

```
class TimeSeries : public TimeSeriesDataTable<double, double>
```

instead of a typedef.

klshrinidhi commented:
@chrisdembia 
* The type aliases have hyperlinks to the respective classes and I have added doxygen documentation referring users to the class documentation. Users click it to go to the right place.
* I have been looking at the doxygen documentation regularly to see if interface looks comprehensible. I think the interface is readable.
* Ajay agreed with you that "TimeSeriesDataTable" is too long but didn't want just "TimeSeries" so I changed it to "TimeSeriesTable" to go with "DataTable". I don't mind changing it again if you have other suggestions.

I think I have made all the changes except:
* Should time column be templated or not ?
* Should we restrict users to add (timestamp, row) pair only to the TimeSeriesTable or also allow them to add rows and timestamps separately with error checking.

aseth1 commented:
@chrisdembia @sherm1 

>Should time column be templated or not ?

I agree that the benefits of time being templated (to support integer or other orderable type) is currently limited and would actually require more work for our Tools and Solvers to use any other type. On the other hand, the proposed implementation already supports different types, so why unwind it? For simplicity and documentation?

In that case, derive `TimeSeriesTable` from `SequentialDataTable<TimeT, ElementT>` and it will be the non-template replacement for Storage today. At the same time we keep the door open to support any sequential data. It would be a shame to undo the capability just to need it again a year from now. You are right that this it introduces complexity for hypothetical benefit, but I see this change being much harder to do a year from now and the currently. I would agree not to go down this route until needed if it wasn't already done.

> Should we restrict users to add (timestamp, row) pair only to the TimeSeriesTable or also allow them to add rows and timestamps separately with error checking

Adding data at a given time (e.g. `addRowAtTime(t, row)`) is convenient/useful and is the predominant way that a user would build a TimeSeriesTable. But currently it is a DataTable, which means a DataAdapter or any other client should be able to `appendRow` and `appendColumn` like it could do with any other DataTable. If we only allow ``addRowAtTime(t, row)` as a way of populating the table, then it isn't a DataTable.  It doesn't have to be, but it currently is.

Data adapters shouldn't have to know about anything other that how to convert data from/to their source and populate/read out a DataTable. If a DataTable meets certain criteria, it should be trivial to convert it to a TimeSeriesTable or to combine a DataTable of "times" with a DataTable of other stuff into a TimeSeriesTable. From this perspective, TimeSeriesTable could construct from and convert to a DataTable for de/serialization but wouldn't be one itself. Then, it could have a more restrictive interface for the user to add rows by time only.

chrisdembia commented:
> For simplicity and documentation?
> 
> In that case, derive TimeSeriesTable from SequentialDataTable<TimeT, ElementT> and it will be the non-template replacement for Storage today. At the same time we keep the door open to support any sequential data. It would be a shame to undo the capability just to need it again a year from now. You are right that this it introduces complexity for hypothetical benefit, but I see this change being much harder to do a year from now and the currently. I would agree not to go down this route until needed if it wasn't already done.

I have mixed feelings about this. One might say that "keeping a feature just because we have it implemented" is one of the worst reasons (even if the implementation is stellar, as in this case), to keep a feature; while simplicity and documentation are great bases for making such choices. I read a blog post about how GitHub has been working on a feature for a few years and the week before they were going to release it they scrapped the whole project because they weren't satisfied with it.

I still think I'm fine with `derive TimeSeriesTable from SequentialDataTable<TimeT, ElementT> `.

> But currently it is a DataTable, which means a DataAdapter or any other client should be able to appendRow and appendColumn like it could do with any other DataTable. If we only allow `addRowAtTime(t, row) as a way of populating the table, then it isn't a DataTable. It doesn't have to be, but it currently is.

If the time column were just another column, then I think we would once again have the invariant that @sherm1 has mentioned. I'm not advocating this; just raising the point. I think there are issues with making time just another column, such as ambiguity about the number of columns. Well, perhaps that ambiguity exists no matter what, but I'd like to think that the number of columns doesn't include the time column. My python code is littered with:

```
for column_name in column_names:
    if column_name != 'time':
        <do calculation>
```

which is very bug prone if someone forgets to exclude the time column.


Thanks for responding to my comments, @klshrinidhi. I'll respond when I get a chance to build and view the doxygen.

chrisdembia commented:
> The type aliases have hyperlinks to the respective classes and I have added doxygen documentation referring users to the class documentation. Users click it to go to the right place.

Okay I just built the doxygen. For Ajay, Sherm, etc., the documentation is completely clear and complete. However, I think the doxygen will intimidate many of our users (for many, OpenSim is how they are learning C++). I would say that there *is* no "right place" for them, because the documentation for `TimeSeriesTable_` has many `template`'s throughout.

This issue is *defnitely* not only with this class; we have it with Sets, and my Delay component, etc.

Perhaps @jimmyDunne could look at the doxygen, since he has a lot of interaction with scripting users.

It could be the case that, for scripting users, we simply  need them to look somewhere else for documentation of TimeSeriesTable.


---

I also think it would be nice to group the methods for TimeSeriesTable, and have an `common accessors and mutators ` and a separate `advanced interface` section.

chrisdembia commented:
To be honest, the documentation does not help a beginner C++ user or scripting user know how to add a row to the TimeSeriesTable. Actually, I myself am intimidated about adding rows. Example code within the doxygen could also help.

In particular, `ArgsToAddRows &&...argsToAddRows` is very confusing. I think it would be very valuable to have an additional method that does not take these optional arguments. I think all methods that take these optional arguments should be in an advanced section.

This is my opinion, and it would be good to hear from @jimmyDunne, @aseth1, @sherm1, @aymanhab, etc.

chrisdembia commented:
@klshrinidhi, I think your written documentation is *very very* clear. My comment is more about the complexity of the interface, and layout of doxygen.

sherm1 commented:
> A TimeSeriesTable is a DataTable so must support `addRow()` etc.

Good point! I forgot about that. I withdraw my comment. That does make me think that TimeSeriesDataTable is a better name than TimeSeriesTable, though.

> I agree that the benefits of time being templated (to support integer or other orderable type) is currently limited and would actually require more work for our Tools and Solvers to use any other type. On the other hand, the proposed implementation already supports different types, so why unwind it? For simplicity and documentation?

This is a bad idea for several reasons: 
- Keeping it because it is there is not a good design principle!
- But besides that, in no sense is the software even close to being done. It is way too early to call it done. It has exactly 0 mileage on it, regression tests remain, we haven't tested it against real use cases like interpolation, we have to document it and explain it to people on the forum, support the API forever.
- I really don't like the idea of introducing an extra layer in the inheritance hierarchy to support this currently-unmotivated feature. As we know from past experience and good API design principles, these hierarchies should be kept as shallow as possible.

Why not have the TimeSeriesTable simply add times to a DataTable? That seems well-motivated, clearly useful, and easy to explain. Then if some other need for row labeling arises, we add another class derived from DataTable, as a sibling of TimeSeriesTable, with and API well-suited to the new application which we will then have a good understanding of. It would be way premature to try to create an abstraction that includes both time and some unknown future use cases. 

Finally, this would be made easier if TimeSeriesTable had a simpler API. As I mentioned in my review comments it currently is very far from a minimally sufficient design and I think that makes it seem as though writing a sibling would be a huge ordeal.

klshrinidhi commented:
@chrisdembia 
Templates are C++'s strength and it is one of the ways we make code highly reusable. C++ standards committee has been and will be adding features to the language and the library to improve and enrich templates. We will be seeing more and more of these in future C++. A C++ programmer needs to understand at-least the basic things. My implementation requires a user to know how to use standard library, for example std::vector. Not write the library itself, but just use it. 
Because the class represents a container, it has to be generic(templated) just like std::vector and as a result many methods have to be templated for flexibility. Most of the overloads of a any function you see in these classes just delegate their work to another overload, which I have mentioned in the documentation so users can see their applicability. 
Once you are back, lets talk so I can separate the commonly used methods and mark the rest as advanced in doxygen.
I will also add some doxygen code examples for functions so they are easier to use.
Also, I will put together better tutorials for users to go through.

klshrinidhi commented:
I am voting to leave the time column templated and constrained to take only arithmetic types.
TimeSeriesTable is generic table holding one generic row per time-step and it should be up to the user to decide what the time-step should be too. For a user, the interface still looks usable with templated time column. It is obvious (to me) when a time is expected as an argument even though it is templated.
For a class intended to be merely  a container, I don't think it adds much complexity.

I will talk to Ajay today, finalize the issues and merge in the changes.

sherm1 commented:
> Templates are C++'s strength and it is one of the ways we make code highly reusable

I agree with this statement, and they are great tools for our own productivity. That doesn't mean though that they are necessarily good when they appear in the visible end-user part of the API, which has to support (a) Python users, (b) Java (GUI, Matlab) users, and (c) relatively inexperienced C++ programmers. Templates there are generally a negative thing because it means there is still work to be done to construct a suitable API for the non-templated languages and easy-to-use-for-beginners C++ examples.

sherm1 commented:
> I am voting to leave the time column templated and constrained to take only arithmetic types

Here are some arithmetic  types: `bool`, `char`, `wchar_t`, `signed char`, `unsigned short`. Why is it a good thing that these types are allowed for the time column?

klshrinidhi commented:
@sherm1 
I don't agree with your comments about templates. 

I am not forcing templates into the API unnecessarily. I have tried to push all the real template work to helper functions sitting under protected access in these classes. These classes, in my opinion, have a rich interface, not a complicated one. For example, there are overloads for a function that take:
* an iterator -- anything that is at-least an input-iterator
* a container -- anything that supports iteration
* a std::initializer list -- proxy for a container that is created by the compiler from user's code

and they all delegate to the function taking an iterator. These overloads have to be templated because they are inherently generic.
Separating some member functions as advanced may make the interface more approachable.

These classes are written to "do one thing really well" and I think it is a good software design principle to follow. They are not trying to do many things at the same time, like the storage class is doing.

Sure, there are problems to be solved regarding bindings for Python/Matlab but I don't think the solution to those problems is to constraint code to work on specific types even though the same code can be re-used for different types. 

Having said, I am okay not having time-column templated. Lets see what Ajay thinks as a final decision.

klshrinidhi commented:
> Here are some arithmetic types: bool, char, wchar_t, signed char, unsigned short. Why is it a good thing that these types are allowed for the time column?

With exception of bool, I think rest can be allowed because time-series is (for me) an ordered collection where order is determined by the time column. As long as the type of the time column supports ordering, I would allow the type.
If it is decided that the time column will be templated, I will disallow bool.

sherm1 commented:
Shrinidhi, I really like the functionality and implementation of the class as a C++ container. If it is only for internal use then it is great. I am really arguing for use cases that incorporate the actual end user of the class. If it is just for C++ programmers, and not intended for use by OpenSim's end user community then the design considerations are different. But if it is intended to support biomechanists' Matlab scripts, then it is important to factor that in to the design. So far I haven't seen any use cases (e.g. for interpolation) so I am having to use my imagination about who the end users of this class are.

klshrinidhi commented:
Ajay and I talked about this last Friday and he suggested writing a non-templated wrapper class for the sake of -- Python, Matlab and C++ programmers who don't know templates yet.

sherm1 commented:
>> Here are some arithmetic types: bool, char, wchar_t, signed char, unsigned short. 

> With exception of bool, I think rest can be allowed

And I presume you are confident that interpolation will still make sense if the time series is instantiated with a wchar_t for time, but I don't (yet) understand the source of that confidence! Isn't that likely to have downstream design implications that ought to be explored first?

klshrinidhi commented:
> And I presume you are confident that interpolation will still make sense if the time series is instantiated with a wchar_t for time, but I don't (yet) understand the source of that confidence! Isn't that likely to have downstream design implications that ought to be explored first?

Ajay and I talked about this too. The class/function doing interpolation should be constrained to work with specific types. There will be a compile time error with appropriate message if doing interpolation on unacceptable types.

sherm1 commented:
> Ajay and I talked about this last Friday and he suggested writing a non-templated wrapper class for the sake of -- Python, Matlab and C++ programmers who don't know templates yet.

That makes sense as a *possible* solution. I hate to see it as an afterthought though -- you are describing above the vast majority of OpenSim users! Also that means the project has a larger scope than we were thinking. That's fine, but I would really like to see this worked through, at least so I don't waste my time and annoy you by making reviewer comments based on a misunderstanding of what you are trying to achieve.

sherm1 commented:
> The class/function doing interpolation should be constrained to work with specific types. There will be a compile time error with appropriate message if doing interpolation on unacceptable types.

In what universe would a *time series* of rows *not* be interpolatable? Even if measured in ints it makes sense to talk about times in between. Time for us is always a continuous quantity. I just don't get why this extra complexity is a good tradeoff. I can see the downside; what is the upside?

klshrinidhi commented:
> In what universe would a time series of rows not be interpolatable? Even if measured in ints it makes sense to talk about times in between. Time for us is always a continuous quantity. I just don't get why this extra complexity is a good tradeoff. I can see the downside; what is the upside?

What types make sense for interpolation should be left to the class doing interpolation. I don't think these containers should be constrained based on that.

aseth1 commented:
@sherm1 @klshrinidhi @chrisdembia let's meet and discuss. How about now?

chrisdembia commented:
I know we seem to have come up with big solutions for the issue that someone could upcast to a DataTable and add a row without a time. 

I wanted to raise the option of using protected or private inheritance, also. I guess this would only solve some of the problem though?

sherm1 commented:
> I wanted to raise the option of using protected or private inheritance

I *think* that would prevent upcasting a TimeSeriesTable to a DataTable which is necessary for passing to a data adapter (not sure though). The method we proposed has some of this flavor -- `DataTable::addRow(rowMetaData, rowData)` would likely be protected, for example.
